client/core/trade.go

```diff
@@ -0,0 +1,655 @@
+	matches struct {
+		maker *msgjson.Match
+		taker *msgjson.Match
+	}
+}
+
+// trackedTrade is an order, its matches, and its cancel order, if applicable.
+// The trackedTrade has methods for handling requests from the DEX to progress
+// match negotiation.
+type trackedTrade struct {
+	order.Order
+	mtx      sync.RWMutex
+	metaData *db.OrderMetaData
+	dc       *dexConnection
+	db       db.DB
+	wallets  *walletSet
+	preImg   order.Preimage
+	sid      string
+	coins    map[string]asset.Coin
+	change   asset.Coin
+	cancel   *trackedCancel
+	matchMtx sync.RWMutex
+	matches  map[order.MatchID]*matchTracker
+}
+
+// newTrackedTrade is a constructor for a trackedTrade.
+func newTrackedTrade(dbOrder *db.MetaOrder, preImg order.Preimage, dc *dexConnection,
+	db db.DB, wallets *walletSet, coins asset.Coins) *trackedTrade {
+
+	ord := dbOrder.Order
+	return &trackedTrade{
+		Order:    ord,
+		metaData: dbOrder.MetaData,
+		dc:       dc,
+		db:       db,
+		wallets:  wallets,
+		preImg:   preImg,
+		sid:      sid(ord.Base(), ord.Quote()),
+		coins:    mapifyCoins(coins),
+		matches:  make(map[order.MatchID]*matchTracker),
+	}
+}
+
+// rate returns the order's rate, or zero if a market or cancel order.
+func (t *trackedTrade) rate() uint64 {
+	if ord, ok := t.Order.(*order.LimitOrder); ok {
+		return ord.Rate
+	}
+	return 0
+}
+
+// coreOrder constructs a *core.Order for the tracked order.Order. If the trade
+// has a cancel order associated with it, the cancel order will be returned,
+// otherwise the second returned *Order will be nil.
+func (t *trackedTrade) coreOrder() (*Order, *Order) {
+	t.mtx.RLock()
+	defer t.mtx.RUnlock()
+	prefix, trade := t.Prefix(), t.Trade()
+	var tif order.TimeInForce
+	if lo, ok := t.Order.(*order.LimitOrder); ok {
+		tif = lo.Force
+	}
+	coreOrder := &Order{
+		Type:        prefix.OrderType,
+		ID:          t.ID().String(),
+		Stamp:       encode.UnixMilliU(prefix.ServerTime),
+		Rate:        t.rate(),
+		Qty:         trade.Quantity,
+		Sell:        trade.Sell,
+		Filled:      trade.Filled,
+		Cancelling:  t.cancel != nil,
+		TimeInForce: tif,
+	}
+	t.matchMtx.RLock()
+	for _, match := range t.matches {
+		dbMatch := match.Match
+		coreOrder.Matches = append(coreOrder.Matches, &Match{
+			MatchID: match.id.String(),
+			Step:    dbMatch.Status,
+			Rate:    dbMatch.Rate,
+			Qty:     dbMatch.Quantity,
+		})
+	}
+	t.matchMtx.RUnlock()
+	var cancelOrder *Order
+	if t.cancel != nil {
+		cancelOrder = &Order{
+			Type:     order.CancelOrderType,
+			Stamp:    encode.UnixMilliU(t.cancel.ServerTime),
+			TargetID: t.cancel.TargetOrderID.String(),
+		}
+	}
+	return coreOrder, cancelOrder
+}
+
+// cancelTrade sets the cancellation data with the order and its preimage.
+func (t *trackedTrade) cancelTrade(co *order.CancelOrder, preImg order.Preimage) {
+	t.mtx.Lock()
+	t.cancel = &trackedCancel{
+		CancelOrder: *co,
+		preImg:      preImg,
+	}
+	t.mtx.Unlock()
+}
+
+// negotiate creates a Match and starts the negotiation thread.
```

Comment:
Since it no longer starts the negotiation thread (not just creates and enters a new `matchTracker`), perhaps a new function name or more precise docs about when and where the negotiation is actually handled (is it `Core.listen`'s ticker)?

Reply:
> is it `Core.listen`'s ticker ?

Actions can be triggered (a `tick`) by any of three events.

1. A request from the server
2. A new block (`tipChange`)
3. The timer in `(*Core).listen` that runs a check 3x per broadcast timeout. See self-review note.

I found that a goroutine for each match wasn't really needed. 

Docs are a little light here. Will improve.